pkg/registry/service/ipallocator/allocator.go

```diff
@@ -0,0 +1,308 @@
+}
+
+// allocateStrategy is a search strategy in the allocation map for a valid IP.
+type allocateStrategy func(allocated *big.Int, max, count int) (int, error)
+
+// NewCIDRRange creates a Range over a net.IPNet.
+func NewCIDRRange(cidr *net.IPNet) *Range {
+	max := RangeSize(cidr)
+	base := bigForIP(cidr.IP)
+	r := Range{
+		net:      cidr,
+		strategy: randomScanStrategy,
+		base:     base.Add(base, big.NewInt(1)), // don't use the network base
+		max:      maximum(0, int(max-2)),        // don't use the network broadcast
+
+		allocated: big.NewInt(0),
+		count:     0,
+	}
+	return &r
+}
+
+func maximum(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// Free returns the count of IP addresses left in the range.
+func (r *Range) Free() int {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.max - r.count
+}
+
+// Allocate attempts to reserve the provided IP. ErrNotInRange or
+// ErrAllocated will be returned if the IP is not valid for this range
+// or has already been reserved.  ErrFull will be returned if there
+// are no addresses left.
+func (r *Range) Allocate(ip net.IP) error {
+	if !r.net.Contains(ip) {
+		return ErrNotInRange
+	}
+
+	offset := calculateIPOffset(r.base, ip)
+	if offset < 0 || offset >= r.max {
+		return ErrNotInRange
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if r.allocated.Bit(offset) == 1 {
+		return ErrAllocated
+	}
+	r.allocated = r.allocated.SetBit(r.allocated, offset, 1)
+	r.count++
+	return nil
+}
+
+// AllocateNext reserves one of the IPs from the pool. ErrFull may
+// be returned if there are no addresses left.
+func (r *Range) AllocateNext() (net.IP, error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	next, err := r.strategy(r.allocated, r.max, r.count)
+	if err != nil {
+		return nil, err
+	}
+	r.count++
+	r.allocated = r.allocated.SetBit(r.allocated, next, 1)
+	return addIPOffset(r.base, next), nil
+}
+
+// Release releases the IP back to the pool. Releasing an
+// unallocated IP or an IP out of the range is a no-op and
+// returns no error.
+func (r *Range) Release(ip net.IP) error {
+	if !r.net.Contains(ip) {
+		return nil
+	}
+
+	offset := calculateIPOffset(r.base, ip)
+	if offset < 0 || offset >= r.max {
+		return nil
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if r.allocated.Bit(offset) == 0 {
+		return nil
+	}
+
+	r.allocated = r.allocated.SetBit(r.allocated, offset, 0)
+	r.count--
+	return nil
+}
+
+// Has returns true if the provided IP is already allocated and a call
+// to Allocate(ip) would fail with ErrAllocated.
+func (r *Range) Has(ip net.IP) bool {
+	if !r.net.Contains(ip) {
+		return false
+	}
+
+	offset := calculateIPOffset(r.base, ip)
+	if offset < 0 || offset >= r.max {
+		return false
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	return r.allocated.Bit(offset) == 1
+}
+
+// Snapshot saves the current state of the pool.
+func (r *Range) Snapshot() (*net.IPNet, []byte) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	return r.net, r.allocated.Bytes()
```

Comment:
We can just switch to RLE and allocate to fill the available space, which means that this can be reduced to on average 4 bytes.  random isn't really useful except to discourage assuming count.  So we can easily allocate in a way to minimize an RLE encoding.

> On May 8, 2015, at 12:37 AM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/registry/service/ipallocator/allocator.go:
> 
> > +	if offset < 0 || offset >= r.max {
> > +		return false
> > +	}
> > +
> > +	r.lock.Lock()
> > +	defer r.lock.Unlock()
> > +
> > +	return r.allocated.Bit(offset) == 1
> > +}
> > +
> > +// Snapshot saves the current state of the pool.
> > +func (r *Range) Snapshot() (*net.IPNet, []byte) {
> > +	r.lock.Lock()
> > +	defer r.lock.Unlock()
> > +
> > +	return r.net, r.allocated.Bytes()
> This means that, if I allocate a /16 for portal_net, this object will be 8 KB. I guess that's not too bad. I hope nobody allocates a /9
> 
> —
> Reply to this email directly or view it on GitHub.
>